[uss_qualifier] NET0260-a performance check for querying flights in an area from a SP

Shastick opened the pull request:
Ensures that the multiple queries to a SP for requesting flights in an area match the required response times.

We have several queries done to that endpoint during the tests already, so this simply extends our aggregate checks.

Observers and service_providers defined in `environment.yaml` (and their respecive Python classes) now have a `uss_base_urls` array parameters that we use to map queries to USS'es. This is required as (to the best of my understanding) we need to be told which urls will be registered for an SP at the DSS.

Note this also includes a little utility for displaying information about a report.

Shastick commented:
Note: not sure if we want to keep the small utility for exploring a report's content? It was helpful to me to learn a bit about how the classes are structured and expect it to help a bit more if we have more complex aggregated checks, but I can also remove it.

mickmis commented:
Before a review, could you ensure the CI passes?
I see a failure in the log, which is unfortunately not caught by the CI, c.f. #162 
Thanks!

Shastick commented:
> Before a review, could you ensure the CI passes? I see a failure in the log, which is unfortunately not caught by the CI, c.f. #162 Thanks!

Looks like the failure was unrelated. Rebased and PR still green and ready for review.

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -65,15 +65,63 @@ def run(self):
+            durations = [query.response.elapsed_s for query in relevant_queries]
+            (p95, p99) = evaluation.compute_percentiles(durations, [95, 99])
+
+            # TODO checks
```
Leftover TODO?

Shastick replied:
My bad, these seem to have vanished into another commit >.> 

Sorry for the confusion will get back to you when I’m sure this PR is good.

Shastick replied:
Wait, the checks are actually there. Need to confirm that the regex matches .

mickmis commented on monitoring/uss_qualifier/report_analyzer.py:
```diff
@@ -0,0 +1,77 @@
+import sys
```
IMO such an utility would be better suited in `monitoring/uss_qualifier/bin` or `monitoring/uss_qualifier/scripts`.

Shastick replied:
Or possibly a `tools` directory either under uss_qualifier or even higher up in the repo?

Shastick replied:
Moved to `scripts` as that directory already exists.

mickmis reviewed (commented):
The check does not seem to validate any query:
```
2023-08-15 12:43:57.628 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss1/flights -> skipped check: no relevant queries
2023-08-15 12:43:57.629 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss2/flights -> skipped check: no relevant queries
```

I assume that is not expected?

Shastick reviewed (commented):
> The check does not seem to validate any query

@mickmis was this for v22a or v19? IIRC I had this passing locally when tested against v22a only.

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -41,13 +43,24 @@ def __init__(
-        self._participants_by_base_url.update(
-            {dp.base_url: dp.participant_id for dp in self._observers}
-        )
+        # identify SPs and observers by their hostnames + ports
```
I believe the comment needs to be updated? It's not hostname+port anymore

mickmis commented on monitoring/uss_qualifier/configurations/dev/library/environment.yaml:
```diff
@@ -44,6 +47,7 @@ net_rid:
             observers:
                 - participant_id: uss2
                   observation_base_url: http://host.docker.internal:8083/riddp/observation
+                  uss_base_urls: [ http://host.docker.internal:8083 ]
```
The `uss_base_urls` is not needed for observers: those URLs correspond to the base URLs that SPs advertise to the DSS. Since DPs/observers don't do such a thing, this is not needed for them.

(note that there is a chance this is needed for the DP aggregate checks, but to be seen at that point)

Shastick replied:
If we don't need this for the observer I'll remove it, can be added back later if required.

mickmis commented on monitoring/uss_qualifier/resources/netrid/observers.py:
```diff
@@ -12,20 +12,39 @@
+    observation_base_url: str
+    """Base URL for the observer's implementation of the interfaces/automated-testing/rid/observation.yaml API"""
+
+    uss_base_urls: List[str]
+    """Base URLS where to reach this particular USS. We allow multiple ones as nothing prevents a
+    USS to advertise multiple different endpoints where to be contacted."""
```
Not needed: c.f. other comment

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -41,13 +43,24 @@ def __init__(
-            {dp.base_url: dp.participant_id for dp in self._observers}
-        )
+        # identify SPs and observers by their hostnames + ports
+        for sp in self._service_providers:
+            for base_url in sp.original_configuration.uss_base_urls:
+                prev_mapping = self._participants_by_base_url.get(base_url)
+                if prev_mapping is not None and prev_mapping != sp.participant_id:
+                    raise ValueError(
+                        f"Invalid configuration detected: same uss base url is defined for mutiple participants. Url: {base_url}, participants: {prev_mapping}, {sp.participant_id}"
+                    )
+                self._participants_by_base_url[base_url] = sp.participant_id
+
+        for obs in self._observers:
+            for base_url in obs.original_configuration.uss_base_urls:
+                prev_mapping = self._participants_by_base_url.get(base_url)
+                if prev_mapping is not None and prev_mapping != obs.participant_id:
+                    raise ValueError(
+                        f"Invalid configuration detected: same uss base url is defined for mutiple participants. Url: {base_url}, participants: {prev_mapping}, {obs.participant_id}"
+                    )
+                self._participants_by_base_url[base_url] = obs.participant_id
```
In both cases (SPs + observers), there is now all `uss_base_urls` correctly mapped, however the `injection_base_url` and `observation_base_url` are not mapped anymore and we need those.

Shastick replied:
Recap of my point during our call: `_participants_by_base_url` is only used in the step below in the constructor to map existing queries to the respective USS.

This is because some queries are made directly to the SP's API (and not via the test harness), such as `http://host.docker.internal:8081/mock/ridsp/v2/uss/flights?view=...`, and we can't rely on the full url of the test harness here. We'd need to read out the hostname, and following our discussion earlier this week that it would in theory be possible for SP's to advertise different host names, I added a separate array to hold these.

My understanding here is that what we care about is mapping any query (to a test/obs harness or the legit API) to the USS they were made to, and for this the hostname part of the URL is enough.

mickmis replied:
> My understanding here is that what we care about is mapping any query (to a test/obs harness or the legit API) to the USS they were made to, and for this the hostname part of the URL is enough.

Nothing prevents the injection API and the SP API of the same USS to have completely different hostnames. 
The change as it is now works for this PR since we are not trying to catch requests made to the injection test harness, but then I find the name of the variable misleading, so IMO either add a comment mentioning that, or include back the injection_base_url in the map.

Shastick replied:
> Nothing prevents the injection API and the SP API of the same USS to have completely different hostnames.

Yes indeed, and in that case they will have several entries in their configuration: the `_participants_by_base_url` map will simply have multiple keys pointing to the same value. Would this cause a problem that I missed?

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -61,19 +74,72 @@ def __init__(
         self.end_test_scenario()
 
+    def _sp_flights_area_times_step(self):
+        pattern = re.compile(r"/flights\?view=")
```
I realize now an issue with the regex (which is also present in the _dp_display_data_times_step I wrote), which is that it is valid to have another parameter after the `?` and before the `view` parameter. The regex should be updated accordingly to reflect this. (and if you don't mind fixing the other one at the same time that would be great)

Shastick replied:
Can do

Shastick replied:
A side question here: does it matter if there are additional parameters being passed? Note that for the queries I am interested in, there actually are.

What might make sense is to give the full method path to avoid accidentally matching something else, but the goal being to find queries to a particular endpoint, I'm not sure there is a need to validate the parameters?

mickmis replied:
> A side question here: does it matter if there are additional parameters being passed? Note that for the queries I am interested in, there actually are more params: Ie http://host.docker.internal:8081/mock/ridsp/v2/uss/flights?view=37.17008162687164%2C-80.60282480412947%2C37.22330680061591%2C-80.5495996303852&recent_positions_duration=60, and there's no guarantee about their order to regex matching too precisely could cause troubles as well.

For the requirement NET0260, no it does not matter.

> What might make sense is to give the full method path to avoid accidentally matching something else, but the goal being to find queries to a particular endpoint, I'm not sure there is a need to validate the parameters?

Indeed you are right, no need to validate the parameters here for NET0260.

mickmis commented on monitoring/uss_qualifier/configurations/dev/library/environment.yaml:
```diff
@@ -17,6 +17,7 @@ net_rid:
             service_providers:
                 - participant_id: uss1
                   injection_base_url: http://host.docker.internal:8071/ridsp/injection
+                  uss_base_urls: [ http://host.docker.internal:8071 ]
```
For the sake of consistency, the base URL should be defined as the whole URL up to path defined in the OpenAPI definition:
```suggestion
                  uss_base_urls: [ http://host.docker.internal:8071/mock/ridsp ]
```
(in the case of the mock USS, this suffix is hard-coded)

mickmis commented on monitoring/uss_qualifier/configurations/dev/library/environment.yaml:
```diff
@@ -26,6 +27,7 @@ net_rid:
             service_providers:
                 - participant_id: uss1
                   injection_base_url: http://host.docker.internal:8081/ridsp/injection
+                  uss_base_urls: [ http://host.docker.internal:8081 ]
```
For the sake of consistency, the base URL should be defined as the whole URL up to path defined in the OpenAPI definition:
```suggestion
                  uss_base_urls: [ http://host.docker.internal:8081/mock/ridsp/v2 ]
```
(in the case of the mock USS, this suffix is hard-coded)

mickmis commented on monitoring/uss_qualifier/resources/netrid/observers.py:
```diff
@@ -12,6 +12,18 @@
 from monitoring.uss_qualifier.resources.communications import AuthAdapterResource
 
 
+class ObserverConfiguration(ImplicitDict):
+    participant_id: str
+    """Participant ID of the observer providing a view of RID data in the system"""
+
+    observation_base_url: str
+    """Base URL for the observer's implementation of the interfaces/automated-testing/rid/observation.yaml API"""
+
+
+class NetRIDObserversSpecification(ImplicitDict):
+    observers: List[ObserverConfiguration]
+
+
```
This change does not appear needed anymore?

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -61,19 +73,72 @@ def __init__(
         self.end_test_scenario()
 
+    def _sp_flights_area_times_step(self):
+        pattern = re.compile(r"v2/uss/flights\?view=")
```
This would catch only v22a queries.

A catch here is that for v19 the OpenAPI specs define `/v1/uss/flights` while for v22a it is `/uss/flights`. So `v2` should not be here in any case.
Also, nice to account for possible other parameters.

Here is a suggestion:
 
```suggestion
        pattern = re.compile(r"(/v1)?/uss/flights\?(\S)*view=")
```

Shastick replied:
We can actually just match on `/uss/flights\?(\S)*view="` I guess

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -119,7 +184,7 @@ def _dp_display_data_times_step(self):
                         severity=Severity.Medium,
                         participants=[participant],
-                        details=f"threshold: {self._rid_version.dp_init_resp_percentile95_s}, 95th percentile: {init_95th}",
+                        details=f"threshold: {self._rid_version.sp_init_resp_percentile95_s}, 95th percentile: {init_95th}",
```
This change does not look right?

mickmis commented:
Additionally, looking at the CI logs something looks wrong. I think it's because of the uss_base_urls having only the hostname and not the rest of the path.

For v22a there is only 1 participant and only the /flights call being evaluated:
```
2023-08-22 05:28:45.802 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:122 - SUCCESS for "ASTM NetRID nominal behavior" scenario
2023-08-22 05:28:46.229 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:104 - Running "ASTM F3411-22a NetRID aggregate checks" scenario...
2023-08-22 05:28:46.229 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: participants -> {'http://host.docker.internal:8081': 'uss2'}
2023-08-22 05:28:46.229 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: nb_queries -> 276
2023-08-22 05:28:46.230 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss2/display_data -> skipped check: no relevant queries
2023-08-22 05:28:46.232 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss2/flights -> percentiles on 37 relevant queries: 95th: 0.034843299999999994; 99th: 0.060277460000000005
2023-08-22 05:28:46.232 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:122 - SUCCESS for "ASTM F3411-22a NetRID aggregate checks" scenario
```

For v19 nothing is evaluated:
```
2023-08-22 05:27:15.854 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:122 - SUCCESS for "ASTM NetRID nominal behavior" scenario
2023-08-22 05:27:16.286 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:104 - Running "ASTM F3411-19 NetRID aggregate checks" scenario...
2023-08-22 05:27:16.287 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: participants -> {'http://host.docker.internal:8071': 'uss2'}
2023-08-22 05:27:16.287 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: nb_queries -> 390
2023-08-22 05:27:16.289 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss2/display_data -> skipped check: no relevant queries
2023-08-22 05:27:16.290 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:263 - Note: uss2/flights -> skipped check: no relevant queries
2023-08-22 05:27:16.290 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:122 - SUCCESS for "ASTM F3411-19 NetRID aggregate checks" scenario
```

Shastick commented:
> Additionally, looking at the CI logs something looks wrong. I think it's because of the uss_base_urls having only the hostname and not the rest of the path.

Indeed, something is off here.

Though we match URLS by prefix for mapping them to participants (via `query.request.url.startswith(base_url)`) so even if we only have the hostname it should still work.

Shastick commented:
Fixed a bug I introduced in `aggregate_checks.py`'s constructor.

We now have, for v19:

```
2023-08-25 14:39:49.132 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: participants_by_base_url -> {'http://v19.ridsp.uss2.localutm/mock/ridsp': 'uss2', 'http://v19.ridsp.uss2.localutm/ridsp/injection': 'uss2', 'http://v19.riddp.uss3.localutm/riddp/observation': 'uss3'}
2023-08-25 14:39:49.133 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss2/display_data -> skipped check: no relevant queries
2023-08-25 14:39:49.135 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss3/display_data -> percentiles on 68 relevant queries (62 different URLs, 66 initial queries, 2 subsequent queries): init 95th: 0.11976845000000001; init 99th: 0.12929327; subsequent 95th: 0.0223759; subsequent 99th: 0.023277580000000003
2023-08-25 14:39:49.137 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss2/flights -> percentiles on 58 relevant queries: 95th: 0.05983205; 99th: 0.06097860000000001
2023-08-25 14:39:49.138 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss3/flights -> skipped check: no relevant queries
```

and for v22a:

```
2023-08-25 14:50:04.666 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: participants_by_base_url -> {'http://v22a.ridsp.uss1.localutm/mock/ridsp/v2': 'uss1', 'http://v22a.ridsp.uss1.localutm/ridsp/injection': 'uss1', 'http://v22a.riddp.uss1.localutm/riddp/observation': 'uss1'}
2023-08-25 14:50:04.668 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss1/display_data -> percentiles on 46 relevant queries (24 different URLs, 31 initial queries, 15 subsequent queries): init 95th: 0.09958079999999998; init 99th: 0.10547775999999999; subsequent 95th: 0.0798888; subsequent 99th: 0.08003536
2023-08-25 14:50:04.670 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss1/flights -> percentiles on 37 relevant queries: 95th: 0.029613199999999996; 99th: 0.03133384
```

`environment.yaml` only configures a `uss1` participant so the discrepancy between v19 and v22a seems fine.

BenjaminPelletier commented on monitoring/uss_qualifier/resources/netrid/service_providers.py:
```diff
@@ -20,6 +20,10 @@ class ServiceProviderConfiguration(ImplicitDict):
     injection_base_url: str
     """Base URL for the Service Provider's implementation of the interfaces/automated-testing/rid/injection.yaml API"""
 
+    uss_base_urls: List[str]
```
I'm hesitant to add an explicit configuration parameter here because I believe the information can be obtained elsewhere, so this introduces the possibility of inconsistency (and additional burden on the people configuring and maintaining tests).  At a high level, it seems reasonable to (at least sometimes) know which participant a given query involves.  So, we could add that information to the [basic Query object](https://github.com/interuss/monitoring/blob/88ad023954e2372deaac1aef2bdc6e142e6433c5/monitoring/monitorlib/fetch/__init__.py#L140):

```python
class Query(ImplicitDict):
    request: RequestDescription
    response: ResponseDescription

    server_id: Optional[str]
    """If specified, identifier of the USS/participant hosting the server involved in this query."""
```

Then, we could populate that information when we discover it.  For direct Service Provider queries, we discover that information [here](https://github.com/interuss/monitoring/blob/88ad023954e2372deaac1aef2bdc6e142e6433c5/monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py#L777), so at that point we can mutate each of `sp_observation.uss_flight_queries` to attach the ParticipantID of the appropriate Service Provider as `server_id`.

Shastick replied:
I'm considering adding that information to the `UTMClientSession` wherever it is possible – I believe most if not all queries are used in conjunction with an instance of that object?

I'm not familiar enough with the codebase to be sure it's a good idea. Will report on my observations.

(Right now I'm weighting off setting the server_id in a more ad-hoc fashion in situations where we know we needed vs trying to cover most/all cases by adding that information at a lower level).

Shastick replied:
After poking around for a bit, I'll start with the ad-hoc approach. `UTMClientSession`'s are instantiated in quite a lot of places and many of them don't have an obvious (to me, anyway) participant id lying around.

Once I have a better overview of the codebase I might reach a different conclusion, but for now I guess that will have to do.

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -41,39 +43,124 @@ def __init__(
         self.end_test_scenario()
 
+    def _sp_flights_area_times_step(self):
+        pattern = re.compile(r"/uss/flights\?(\S)*view=")
```
Similar to the above comment relating to annotating information during the test rather than trying to reconstruct it afterward, we could annotate queries with the type of query they are; e.g.:

```python
class QueryType(str, Enum):
    F3411v22aFlights = "astm.f3411.v22a.sp.flights"
    F3411v19Flights = "astm.f3411.v19.sp.flights"
    F3411v22aFlightDetails = "astm.f3411.v22a.sp.flight_details"


class Query(ImplicitDict):
    request: RequestDescription
    response: ResponseDescription

    query_type: Optional[QueryType]
    """If specified, the recognized type of this query."""
```

The `query_type` above would presumably be populated, e.g., [here](https://github.com/interuss/monitoring/blob/88ad023954e2372deaac1aef2bdc6e142e6433c5/monitoring/monitorlib/fetch/rid.py#L752C32-L752C32).

Shastick commented:
Thanks for the feedback, @BenjaminPelletier, both suggestions seem doable to me, I'll get back to you if I need any clarification.

Shastick commented:
Taking this to a draft state again until it's in a satisfactory state

Shastick commented:
Ready for another pass of review. The current version is working, as far as I can trust the logs:

```
2023-08-31 20:04:52.192 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:104 - Running "ASTM F3411-22a NetRID aggregate checks" scenario...
2023-08-31 20:04:52.193 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: participants -> {'http://v22a.ridsp.uss1.localutm/ridsp/injection': 'uss1', 'http://v22a.riddp.uss1.localutm/riddp/observation': 'uss1'}
2023-08-31 20:04:52.193 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: nb_queries -> 290
2023-08-31 20:04:52.193 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: service_providers -> configured service providers: uss1 - http://v22a.ridsp.uss1.localutm/ridsp/injection
2023-08-31 20:04:52.193 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: observer -> configured observer: uss1 - http://v22a.riddp.uss1.localutm/riddp/observation
2023-08-31 20:04:52.196 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss1/display_data -> percentiles on 46 relevant queries (24 different URLs, 34 initial queries, 12 subsequent queries): init 95th: 0.07079325; init 99th: 0.07741384999999999; subsequent 95th: 0.0712499; subsequent 99th: 0.07249478000000001
2023-08-31 20:04:52.197 | INFO     | monitoring.uss_qualifier.scenarios.scenario:record_note:270 - Note: uss1/flights -> percentiles on 56 relevant queries: 95th: 0.027803249999999998; 99th: 0.02819827
2023-08-31 20:04:52.197 | INFO     | monitoring.uss_qualifier.suites.suite:_run_test_scenario:122 - SUCCESS for "ASTM F3411-22a NetRID aggregate checks" scenario
```

v19 also has output showing that queries are properly analysed.

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:
```diff
@@ -234,6 +234,9 @@ def evaluate_system_instantaneously(
                 self._test_scenario.record_query(q)
+                # Keep track of who the query is being made to
+                # for later use in aggregate checks
+                q.server_id = self._dss.participant_id
```
`sp_observation` is a [`FetchedFlights`](https://github.com/interuss/monitoring/blob/0da60538c27675aebde2b1826a155dd163927465/monitoring/monitorlib/fetch/rid.py#L896) which consists of a [query to the DSS](https://github.com/interuss/monitoring/blob/0da60538c27675aebde2b1826a155dd163927465/monitoring/monitorlib/fetch/rid.py#L897), [queries to various USSs for flights](https://github.com/interuss/monitoring/blob/0da60538c27675aebde2b1826a155dd163927465/monitoring/monitorlib/fetch/rid.py#L898), and [queries to various USSs for flight details](https://github.com/interuss/monitoring/blob/0da60538c27675aebde2b1826a155dd163927465/monitoring/monitorlib/fetch/rid.py#L899).  The DSS participant ID only applies to the first of those, so we wouldn't want to apply it to all the USS queries as well.

At this point, we don't know which of [`sp_observation.uss_flight_queries`](https://github.com/interuss/monitoring/blob/0da60538c27675aebde2b1826a155dd163927465/monitoring/monitorlib/fetch/rid.py#L898) belongs to which (if any) of the test participants -- it's conceivable that another USS not participating in the test may nonetheless be in the interoperability ecosystem, causing uss_qualifier and other USSs to interact with it during the test.  We only deduce which interoperability ecosystem SPs must be which test participants [here](https://github.com/interuss/monitoring/blob/88ad023954e2372deaac1aef2bdc6e142e6433c5/monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py#L777) when we map the observed flights to injected flights.  Since (at that point) we know which participant each injected flight was sent to, we can, at this point, associate observed flights with test participants.

Shastick replied:
Ah, indeed. 

Stamping the query with the id now happens after the mapping was established.

Shastick commented on monitoring/uss_qualifier/resources/netrid/service_providers.py:
```diff
@@ -38,16 +38,19 @@ class NetRIDServiceProvider(object):
     participant_id: str
     base_url: str
     client: infrastructure.UTMClientSession
+    original_configuration: ServiceProviderConfiguration
```
No indeed, will remove.
